Typescript React Recommendations (#29)
strimzi/strimzi-ui: docs/Coding.md

#Coding Standards

> The strmzi UI projects uses best practices based off the offical [React TypeScript Cheatsheet](https://react-typescript-cheatsheet.netlify.app/), with modificiations for this project. The React TypeScript Cheatsheet is maintained and used by developers through out the world, and is a place where developers can bring togehter lessons learned using TypeScript and React.

Reviewer: Typo 
```suggestion
The strimzi UI project uses best practices based off the official [React TypeScript Cheatsheet](https://react-typescript-cheatsheet.netlify.app/), with modifications for this project. The React TypeScript Cheatsheet is maintained and used by developers through out the world, and is a place where developers can bring together lessons learned using TypeScript and React.
```

Author: Fixed the typos thanks!

## Imports

Since we are using TypeScript 4.x + for this project, default imports should conform to the new standard set forth in TypeScript 2.7:

```javascript
import React from 'react';
import ReactDOM from 'react-dom';
```

For imports that are not the default import use the following syntax:

```javascript
import { X1, X2, ... Xn } from 'package-x';
```

## View Props

> Since strmizi UI is made up of 2 different views (one for PatternFly, and one for Carbon) props for a component should be declared in seperate file instead of within the components source code. For example Carbon views are written in **View.carbon.tsx**, while PatternFly views are written in **View.patternfly.tsx**. In order to allow the views to share the same props, the code for the props should live in **View.props.tsx**.

Reviewer: ```suggestion
Since the strimzi UI is made up of 2 different views (one for PatternFly, and one for Carbon), props for a component should be declared in a separate file instead of within the components source code. For example Carbon views are written in **View.carbon.tsx**, while PatternFly views are written in **View.patternfly.tsx**. In order to allow the views to share the same props, the code for the props should live in **View.props.tsx**.
```

> For props we are using **type** instead of **interfaces**. The reason to use types instead of interfaces is for consistency betweeen the views and because it's more constrained (See [Types or Interfaces](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/types_or_interfaces) for more clarifiation). By using types we are ensuring that both views will not deviate from the agreed upon [contract](https://dev.to/reyronald/typescript-types-or-interfaces-for-react-component-props-1408).

Reviewer 2: Does this mean we're not using `prop-types`? If so, are default props defined in the component function signature or elsewhere?

Author: For typescript we don't need prop types as they take they are replace by type definition.  I will add documentation and example of how to set the default props.

Author: Added an example of how to set default props.

The following is an example of using a type for props:

```javascript
import React, { ReactNode } from 'react'

...

export type ExampleComponentProps = {
    message: string;
    children: ReactNode;
}
```

> ## Maintiaing state with types

Reviewer: ```suggestion
## Maintaining state with types
```

Author: Fixed

> When maintaining state for a component it is reccommend that you use a [type instead of an interface](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/types_or_interfaces).

Reviewer 2: `reccomend` -> `recommended`

Author: Reworded this section

```javascript
type ApiDrawerState = {
  currentApiId: string,
  isExpanded: boolean,
};
```

> ## Intefaces

Reviewer 2: typo

Author: done.

> ## Intefaces

Reviewer: ```suggestion
## Interfaces
```

Author: Fixed

> Interfaces should be used for all public facing API definitions. A table describing when to use interfaces vs. types can be found [here](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/types_or_interfaces).

Reviewer: Example? When would expect to use them?

## Function Components

> This project uses function components and hooks over class components. When coding function components in typescript a developer should include any specific props from the **View.props.tsx**

Reviewer: Example import?

Author: I removed the imports

```javascript
import React, { FunctionComponent } from 'react';

...

export const ExampleComponent: FunctionComponent<ExampleComponentProps> = ({ message, children }: ExampleComponentProps) => (
  <ReactFragment>
    <div>{message}</div>
    <div>{children}</div>
  </<ReactFragment>>
);
```

> For components that do not have any additonal props an empty object should be used instead:

Reviewer: ```suggestion
For components that do not have any additional props an empty object should be used instead:
```

```javascript
import React, { FunctionComponent } from 'react';

...

export const ExampleNoPropsComponent: FunctionComponent<{}> = () => (
  <div>Example Component with no props</div>
);
```

Additonal details around function components can be found [here](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/function_components).

## Hooks

When using hooks with Typescript there are few recommendations that we follow below. Additional recommendations besides the ones mention in this document can be found [here](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks).

### Inference vs Types for useState

Currently we recommend using inference for the primative types booleans, numbers, and string when using useState. Anything other then these 3 types should use a declartive syntax to specify what is expected. For example the following is an example of how to use inference:

```javascript
const [isEnabled, setIsEnabled] = React.useState(false);
```

Here is an example how to use a declaritive syntax. When using a declearative syntax if the value can be null that will also need to be specified:

```javascript

> const [user, setUser] = React.useState<IUser | null>(null);

Reviewer: Do these need importing?

Author: We can do it either way.  I'll change it to import useState directly

...

setUser(newUser);

```

> ### useReducres

Reviewer 2: typo

Author: done

> ### useReducres

Reviewer: ```suggestion
### useReducers
```

> When using reducers make sure you specify the [return type and not use inference](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks#usereducer).

Reviewer: ```suggestion
When using reducers make sure you specify the [return type and do not use inference](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks#usereducer).
```

> When using reducers make sure you specify the [return type and not use inference](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks#usereducer).

Reviewer: Example? Saves having to leave page

### useEffect

> For useEffect only [return the function or undefined](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks#useeffect).

Reviewer: Example? Saves having to leave page

### useRef

When using useRef there are two options with Typescript. The first one is when creating a [read-only ref](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks#useref).

```javascript
const refExample = useRef<HTMLElement>(null!);
```

> By passing in null! it will prevent Typescript from returning an error saying refExample maybe null.

Reviewer 2: Less of a review comment but a typescript questions. How much should the `!` operator be avoided? It seems quite dangerous

Reviewer: Wonder if there's an ESLint rule that complains if it's used other than some cases?

Author: I would try not to use it except in this case.  I can check to see if there's a better solution.

> The second option is for createing [mutable refs](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks#useref) that you will manage.

Reviewer: ```suggestion
The second option is for creating [mutable refs](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks#useref) that you will manage.
```

Author: fixed

#Coding Standards

The Strimzi UI projects uses best practices based off the official [React TypeScript Cheat sheet](https://react-typescript-cheatsheet.netlify.app/), with modifications for this project. The React TypeScript Cheat sheet is maintained and used by developers through out the world, and is a place where developers can bring together lessons learned using TypeScript and React.

## Imports

Since we are using TypeScript 4.x + for this project, default imports should conform to the new standard set forth in TypeScript 2.7:

> ```javascript

Reviewer 3: javascript -> typescript ?

Reviewer 3: This comment applies for all snippets here - could they all be updated?

Author: yes should be typescript.

```javascript
import { X1, X2, ... Xn } from 'package-x';
```

## View Props

Since Strimzi UI is made up of 2 different views (one for PatternFly, and one for Carbon) props for a component should be declared in separate file instead of within the components source code. For example Carbon views are written in **View.carbon.tsx**, while PatternFly views are written in **View.patternfly.tsx**. In order to allow the views to share the same props, the code for the props should live in **View.props.tsx**.

For props we are using **type** instead of **interfaces**. The reason to use types instead of interfaces is for consistency between the views and because it's more constrained (See [Types or Interfaces](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/types_or_interfaces) for more clarification). By using types we are ensuring that both views will not deviate from the agreed upon [contract](https://dev.to/reyronald/typescript-types-or-interfaces-for-react-component-props-1408).

The following is an example of using a type for props:

```javascript
import React, { ReactNode } from 'react'

...

> export type ExampleComponentProps = {

Reviewer: If I've understood https://github.com/DefinitelyTyped/DefinitelyTyped/blob/23a1fd68ce20cb34d944903c44c58b3897fcfd19/types/react/index.d.ts#L546 correctly, you won't need to add children in the props as they'll be injected when you use `FunctionComponent`

Author: I'll remove it.
